Add compaction API

jowl opened the pull request:
This PR suggests adding methods from the compactions API. Specifically, `#compact_range` which forces an immediate compaction, optionally of a specific range, and `#full_compaction` which will compact the whole range until there are no changes in any sstables.

For some reason `#compact_range` doesn't do a "full" compaction, and subsequent calls might compact further, why I've added the `#full_compaction` helper. It compares the value of the `leveldb.sstables` property from before and after a compaction, and reruns when necessary. According to a comment in the source code, the property holds "a multi-line string that describes all of the sstables that make up the db contents", so the comparison is a bit heavy if the number of sstables is large. But then a compaction is as well.

I've also locked RSpec to `~> 2.99` as some tests use syntax that isn't supported in newer versions of RSpec.

EDIT: Removed `#suspend_compactions` and `#resume_compactions` and added `#full_compaction`

Commit:
Add #suspend/resume_compactions

Commit:
Add #compact_range for manual compactions

Commit:
Lock rspec to ~> 2.99

As some of the tests are not working in RSpec 3.0

Commit:
Add compaction docs

jowl commented on spec/level_db_spec.rb:
```diff
@@ -65,6 +65,50 @@
+      it "is ok to omit all options" do
+        LevelDb::Db.new(leveldb).compact_range
+        leveldb.should have_received(:compact_range).with(nil, nil)
+      end
```
I really don't like these tests, but I couldn't really think of any way to test the compactions.

iconara replied:
I think they're fine given the circumstances.

iconara commented:
LGTM, but fix the tests (byte arrays are not equal, at least not on Travis). I've given you commit rights and added you as owner of the gem. Fix the tests and release this as v1.1, your choice. Don't forget to [add a release](https://github.com/iconara/leveldb-jruby/releases).

iconara commented on spec/level_db_spec.rb:
```diff
@@ -65,6 +65,50 @@
+        leveldb.should have_received(:compact_range).with('one'.to_java_bytes, 'three'.to_java_bytes)
+      end
+
+      it "is ok to omit the 'from' option" do
```
Maybe change this to 'compacts the whole range from the given start byte when to is omitted' (and similar for the other tests). It's better to say what happens when you omit the option than saying that it's ok to omit it.

jowl replied:
Good point. I'll change the descriptions.

Commit:
Update compaction tests

Commit:
Remove #suspend/resume_compactions again

After reading the source, it appears that the features in the underlying
library aren't really stable why I've opted to hide these features again.

Commit:
Add #full_compaction

For some reason, it appears that a call to #compact_range doesn't do the
trick. This commit adds a helper method that inspects the sstables
before and after the compaction and reruns the compaction when necessary
to determine when the db is maximally compacted.

jowl commented:
After reading through the sources of the leveldb deps with @grddev, we found that `#suspend_compactions` and `#resume_compactions` wasn't really well-behaved and I've thus chosen to remove them from this PR.

I also found that `#compact_range` doesn't always finish the job, and have added a helper method (`#full_compaction`) which calls `#compact_range` until the db is maximally compacted.

I've updated the PR description accordingly. Could you please take another look @iconara?

iconara commented on lib/level_db.rb:
```diff
@@ -113,6 +113,24 @@ def each(options={}, &block)
+        pre_compaction_sstables = sstables
+        compact_range
+        sstables = @db.get_property(SSTABLES_PROPERTY)
+      end
```
Are you sure this can't end up in an infinite loop? What if other threads manipulate the database at the same time? Or would that just make this loop go another go, and then stop?

jowl replied:
Thank you for noticing, I totally forgot about this. My original intention was to add a maximum iterations options which would terminate the compactions regardless of convergence. I'll revise it.

And also, there are other properties called `leveldb.num-files-at-level<N>` where `N` is an integer. Do you think it'd be nicer/more clear to have a termination condition based on the number of files? It is a bit cumbersome to count them all, since there's no way of knowing how many levels there are, so you'd have to do something like

``` ruby
count = i = 0
while (value = @db.get_property("leveldb.num-files-at-level#{i}"))
  count += value.to_i
  i += 1
end
count
```

My main concern about the current solution is that I'm not sure about the contents of the `leveldb.sstables` property, and specifically its sorting. It's just a multi-line string that describes all sstables, but I don't know if its order is deterministic. I'm guessing that it is, but I don't know. And it feels like an unstable termination condition to make a string comparison like that when the format isn't well documented.

iconara replied:
Sounds like there are quite a few problems with this loop. It would be reasonable to assume that the order of the files in the `leveldb.sstables` property would be the same from call to call, but _reasonable_ is probably not enough. I wouldn't be comfortably with anything less than sorting the lists before comparing (I didn't even think of this on my first look at the code, so good that you pointed that out).

As for the level counting I don't know if that's better or worse. What if the same number of SSTables were added that were compacted away between two calls? I say that unless there's anything else wrong with looking at the files then there should be no problem looking at the list of files (provided it's sorted).

A max iterations parameter sounds like a must too. I would raise an error if there was no convergence after the max number of iterations.

grddev replied:
Looking at the leveldb source code, [the maximum number of levels seems to be 7](https://github.com/google/leveldb/blob/77948e7eec0613fb5cbecb7e320b9498607030b5/db/dbformat.h#L22). Also, [the format of returned string doesn't strike me as easily parsed/sorted](https://github.com/google/leveldb/blob/77948e7eec0613fb5cbecb7e320b9498607030b5/db/version_set.cc#L565-L589).

jowl replied:
But then I should be able to compare the number of files per level like

``` ruby
def num_files_per_level
  7.times.map { |i| @db.get_property("leveldb.num-files-at-level#{i}") }
end

pre_compaction = num_files_per_level
compact_range
post_compaction = num_files_per_level
converged = post_compaction == pre_compaction
```

That should be "safe", right? We'll obviously keep going if we're writing to the DB while compacting, but that should be true regardless of which method we use. But using this method, at least we won't be exposed to

> the same number of SSTables were added that were compacted away between two calls

if I'm not mistaken?

iconara replied:
I prefer the variant that didn't assume a fixed number of levels. It may change, and it could be different between say LevelDB and RocksDB.

iconara replied:
But I guess it should be safe. I can see how a pathological combination of events could lead it to loop for a long time, but it should converge eventually.

iconara commented on spec/level_db_spec.rb:
```diff
@@ -65,6 +65,62 @@
+        calls.should include(['one', 'three'])
+      end
+
+      it "compacts the whole range to the 'to' key when the 'from' option is omitted" do
```
Why not `it 'compacts … "to" … "from" … '`? (I'm thinking of quoting consistency).

jowl replied:
I like consistency. I'll change it.

iconara commented:
Too bad about the suspend/resume thing, but better to leave them out than keep them in. I've made two comments, but otherwise LGTM.

Commit:
Raise ConvergenceError in #full_compaction

Commit:
Quoting consistency

Commit:
Look at num-files-at-level<L> rather than sstables

when determining compaction convergence
